bigtable/internal/testproxy/proxy_test.go

```diff
@@ -0,0 +1,448 @@
+	}
+	defer dataClient.Close()
+
+	t := dataClient.Open(tbl)
+
+	for fc := 0; fc < count; fc++ {
+		for cc := count; cc > 0; cc-- {
+			for tc := 0; tc < count; tc++ {
+				rmw := bigtable.NewReadModifyWrite()
+				rmw.AppendValue(fmt.Sprintf("%s%d", cf, fc), fmt.Sprintf("coll%d", cc), []byte("test data"))
+
+				_, err = t.ApplyReadModifyWrite(ctx, rk, rmw)
+				if err != nil {
+					return fmt.Errorf("testproxy setup: failure populating row: %v", err)
+				}
+			}
+		}
+	}
+
+	return nil
+}
+
+/*
+TestMain has three threads that it needs to start:
+1) The mocked Bigtable service (server)
+2) The NewCloudBigtableV2TestProxyClient client that sends requests to the testproxy server.
+3) The testproxy server under test
+
+	The communication sequence looks kind of like:
+
+	TestProxyClient <=> test proxy server (what we're testing) <=> mocked BT server
+*/
+func TestMain(m *testing.M) {
+	ctx := context.Background()
+
+	// 1) Start the mocked Bigtable service
+	// This requires creating a "table" in memory
+	bts, err := bttest.NewServer(bta)
+	if err != nil {
+		log.Fatalf("testproxy setup: can't create inmem Bigtable server")
+	}
+	err = populateTable(bts)
+	if err != nil {
+		log.Fatalf("testproxy setup: can't populate mock table")
+	}
+
+	// 3) Start the test proxy server
+	lis = bufconn.Listen(buf)
+	s := newProxyServer(lis)
+	go func() {
+		if err := s.Serve(lis); err != nil {
+			log.Fatalf("failed to serve: %v", err)
+		}
+	}()
+
+	// 2) Create the test proxy client
+	conn2, err := grpc.DialContext(ctx, lis.Addr().String(), grpc.WithContextDialer(bufDialer), grpc.WithTransportCredentials(insecure.NewCredentials()))
+	if err != nil {
+		log.Fatalf("testproxy setup: failed to dial testproxy: %v", err)
+	}
+	defer conn2.Close()
+	c := pb.NewCloudBigtableV2TestProxyClient(conn2)
+	client = &c
+
+	// This could create a little bit of a race condition with the previous
+	// go routine ...
+	// However, we need to have the test proxy server running in order to create
+	// a reusable client stored in its memory
+	req := &pb.CreateClientRequest{
+		ClientId:   tpc,
+		ProjectId:  "client",
+		DataTarget: bta,
+		InstanceId: "instance",
+	}
+
+	_, err = (*client).CreateClient(ctx, req)
+	if err != nil {
+		log.Fatalf("testproxy setup:  CreateClient() failed: %v", err)
+	}
+
+	m.Run()
+}
+
+func TestCreateClient(t *testing.T) {
+	// Test
+	cid := "testCreateClient"
+	ctx := context.Background()
+
+	req := &pb.CreateClientRequest{
+		ClientId:   cid,
+		ProjectId:  "client",
+		DataTarget: bta,
+		InstanceId: "instance",
+	}
+
+	_, err := (*client).CreateClient(ctx, req)
+	if err != nil {
+		t.Fatalf("testproxy test: CreateClient() failed: %v", err)
+	}
+
+	// Teardown
+	_, err = (*client).RemoveClient(ctx, &pb.RemoveClientRequest{
+		ClientId:  cid,
+		CancelAll: true,
+	})
+
+	if err != nil {
+		t.Errorf("testproxy test: CreateClient() teardown failed: %v", err)
+	}
+}
+
+func TestRemoveClient(t *testing.T) {
```

Comment:
Is this a duplicate of the test above? Shall it be rolled into the test above?

Reply:
Yeah, `TestCreateClient()` and `TestRemoveClient()` are essentially two halves of the same coin. They should be combined.

Reply:
Merged / remove duplicate